File: camic/lib/foundation/diag.ml
(* source, lexical, and syntax error will cause the abort of compilation *)
exception AbortCompilation of string * int

let source ?pos msg =
  match pos with
  | None -> raise @@ AbortCompilation (("[Source Error]" ^ msg), -1)
  | Some (p : Token.position) -> raise @@ AbortCompilation ((Printf.sprintf "%s:%d:%d: [Encoding Error]%s" p.file p.line p.column msg), -1)

let lex_pps line ?column ?file msg =
  let source_name = Option.value ~default:(if !Config.source_name = "" then "<stdin>" else !Config.source_name) file in
  let column_info = match column with None -> "" | Some v -> Printf.sprintf ":%d" v in
  raise @@ AbortCompilation ((Printf.sprintf "%s:%d%s: [LexPreprocess Error]%s" source_name line column_info msg), -2)

let lexical (pos : Token.position) msg = raise @@ AbortCompilation ((Printf.sprintf "%s:%d:%d: [Lexical Error]%s" pos.file pos.line pos.column msg), -3)
let parsing (pos : Token.position) msg = raise @@ AbortCompilation ((Printf.sprintf "%s:%d:%d: [Parsing Error]%s" pos.file pos.line pos.column msg), -4)

module Warning = struct
  let lexical (pos : Token.position) msg = Printf.eprintf "\x1b[033m%s:%d:%d: [Lexical Warning]%s\x1b[0m\n" pos.file pos.line pos.column msg
end

File: camic/src/Foundation/Diag.hs
module Foundation.Diag
  ( module Foundation.Diag,
  )
where

import Config (Config (sourceName), config)
import Control.Exception (Exception, throw)
import Data.Data (Typeable)
import Foundation.Loc (Loc)
import Text.Printf (printf)

--  source, lexical, and syntax error will cause the abort of compilation
data AbortCompilation = AbortCompilation String Int deriving (Show, Typeable)

instance Exception AbortCompilation

systemError :: String -> a
systemError msg = throw $ AbortCompilation msg (-1)

sourceError :: String -> Maybe Loc -> a
sourceError msg Nothing = throw $ AbortCompilation msg (-2)
sourceError msg (Just loc) = throw $ AbortCompilation (printf "%s: [Encoding Error]%s" (show loc) msg) (-2)

lexppsError :: Int -> Maybe String -> Maybe String -> String -> a
lexppsError line column file msg =
  let sourceName' = maybe (if config.sourceName == "" then "<stdin>" else config.sourceName) id file
      columInfo = maybe "" (printf ":%d" :: String -> String) column
   in throw $ AbortCompilation (printf "%s:%d%s: [LexPreprocess Error]%s" sourceName' line columInfo msg) (-3)

lexicalError :: Loc -> String -> a
lexicalError loc msg = throw $ AbortCompilation (printf "%s: [LexicalError]%s" (show loc) msg) (-3)

parsingError :: Loc -> String -> a
parsingError loc msg = throw $ AbortCompilation (printf "%s: [ParsingError]%s" (show loc) msg) (-4)

lexicalWarning :: Loc -> String -> String
lexicalWarning loc msg = (printf "%s: [LexicalWarning]%s" (show loc) msg)
